src/algebra/group/hom.lean

```diff
@@ -184,15 +243,199 @@ attribute [instance] inv.is_group_hom neg.is_add_group_hom
+mk'' (λ g, (f g)⁻¹) $ λ a b, by rw [←mul_inv, f.map_mul]
+
+end monoid_hom
+
+namespace add_monoid_hom
+variables {A : Type*} {B : Type*} {C : Type*} [add_monoid A] [add_monoid B] [add_monoid C]
+{G : Type*} [add_group G] {H : Type*} [add_group H]
+
+/-- two additive monoid homomorphisms with equal underlying maps are equal-/
+@[extensionality] def ext (f g : A →+ B) (h : (f : A → B) = g) : f = g :=
+by cases f; cases g; cases h; refl
+
+attribute [to_additive add_monoid_hom.ext] monoid_hom.ext
+attribute [to_additive add_monoid_hom.ext.equations._eqn_1] monoid_hom.ext.equations._eqn_1
+
+/-- if f is a additive monoid homomorphism then f 0 = 0-/
+@[simp] def map_zero (f : A →+ B) : f 0 = 0 := f.map_zero'
+
+attribute [to_additive add_monoid_hom.map_zero] monoid_hom.map_one
+
+/-- if f is an additive monoid homomorphism then f (a + b) = f a + f b -/
+@[simp] def map_add (f : A →+ B) : ∀ a b : A, f (a + b) = f a + f b := f.map_add'
+
+attribute [to_additive add_monoid_hom.map_add] monoid_hom.map_mul
+
+/-- the identity map from an add_monoid to itself -/
+def id (A : Type*) [add_monoid A] : A →+ A :=
+{ to_fun := id,
+  map_zero' := rfl,
+  map_add' := λ _ _, rfl}
+
+attribute [to_additive add_monoid_hom.id._proof_1] monoid_hom.id._proof_1
+attribute [to_additive add_monoid_hom.id._proof_2] monoid_hom.id._proof_2
+attribute [to_additive add_monoid_hom.id.equations._eqn_1] monoid_hom.id.equations._eqn_1
+
+/-- composition of additive monoid morphisms is an additive monoid morphism -/
+def comp (fbc : B →+ C) (fab : A →+ B) : A →+ C :=
+{ to_fun := fbc ∘ fab,
+  map_zero' := by simp,
+  map_add' := by simp,
+}
+
+attribute [to_additive add_monoid_hom.comp] monoid_hom.comp
+attribute [to_additive add_monoid_hom.comp._proof_1] monoid_hom.comp._proof_1
+attribute [to_additive add_monoid_hom.comp._proof_2] monoid_hom.comp._proof_2
+attribute [to_additive add_monoid_hom.comp.equations._eqn_1] monoid_hom.comp.equations._eqn_1
+
+/-- The sum of two additive monoid morphisms is an additive monoid morphism if the
+target is commutative-/
+def add {A B} [add_monoid A] [add_comm_monoid B] (f g : A →+ B) : A →+ B :=
+{ to_fun := λ a, f a + g a,
+  map_zero' := show f 0 + g 0 = 0, by simp,
+  map_add' := begin intros, show f (x + y) + g (x + y) = f x + g x + (f y + g y),
+    rw [f.map_add, g.map_add, ←add_assoc, ←add_assoc, add_right_comm (f x)], end
+}
+
+attribute [to_additive add_monoid_hom.add] monoid_hom.mul
+attribute [to_additive add_monoid_hom.add._proof_1] monoid_hom.mul._proof_1
+attribute [to_additive add_monoid_hom.add._proof_2] monoid_hom.mul._proof_2
+attribute [to_additive add_monoid_hom.add.equations._eqn_1] monoid_hom.mul.equations._eqn_1
+
+/-- additive group homomorphisms preserve additive inverse -/
+theorem map_neg (f : G →+ H) (a : G) : f (-a) = -(f a) :=
+eq_neg_of_add_eq_zero $ by rw [←f.map_add, neg_add_self, f.map_zero]
+attribute [to_additive add_monoid_hom.map_neg] monoid_hom.map_inv
+
```

Comment:
Maybe also add the simp-lemma `map_sub`?

Reply:
OK!